File: python/day2/main.py
import itertools
import functools
import re
import math

from copy import deepcopy

INPUT_FILE = 'input.txt'

def read_file(filename, func=None):
    result = []
    with open(filename, 'r') as f:
        for line in f:
            result.append(func(line.strip()) if func else line.strip())
    return result

def main1():
    lines = read_file(INPUT_FILE)
    position = 0
    depth = 0
    for l in lines:
        inst, value = l.split(' ')
        value = int(value)
        if inst == "forward":
            position += value
        elif inst == "down":
            depth += value
        else:
            depth -= value
    print(position * depth)

def main2():
    lines = read_file(INPUT_FILE)
    position = 0
    depth = 0
    aim = 0
    for l in lines:
        inst, value = l.split(' ')
        if inst == "forward":
            position += int(value)
            depth += aim * int(value)
        elif inst == "down":
            aim += int(value)
        else:
            aim -= int(value)
    print(depth * position)
    
if __name__ == "__main__":
    main1()
    main2()

File: rust/day2/src/main.rs
use day2::util;

fn main() {
    let lines = util::read_lines("input.txt").unwrap();
    part1(&lines);
    part2(&lines);
}

fn part1(lines: &Vec<String>) {
    let mut position = 0;
    let mut depth = 0;
    for l in lines {
        if let Some((instruction, value)) = l.split_once(" ") {
            let int_value = value.parse::<i32>().unwrap();
            match instruction {
                "forward" => position += int_value,
                "down" => depth += int_value,
                "up" => depth -= int_value,
                _ => {}
            }
        }

    }
    println!("{}", position * depth);
}

fn part2(lines: &Vec<String>) {
    let mut position = 0;
    let mut depth = 0;
    let mut aim = 0;
    for l in lines {
        if let Some((instruction, value)) = l.split_once(" ") {
            let i_value = value.parse::<i32>().unwrap();
            match instruction {
                "forward" => {
                    position += i_value;
                    depth += aim * i_value
                },
                "down" => aim += i_value,
                "up" => aim -= i_value,
                _ => {}
            }
        }

    }
    println!("{}", position * depth);
}
